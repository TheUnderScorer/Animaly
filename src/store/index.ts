import { applyMiddleware, combineReducers, createStore } from 'redux';
import { UserActions, userReducer, UserState } from './reducers/user/reducer';
import thunk from 'redux-thunk';
import { initialState as userInitialState } from './reducers/user/reducer';

export interface AppStore {
  user: UserState;
}

const initialState: AppStore = {
  user: userInitialState,
};

export type AppActions = UserActions;

const reducers = combineReducers({
  user: userReducer,
});
const middleware = [thunk];

const store = createStore(
  reducers,
  initialState,
  applyMiddleware(...middleware),
);

export default store;
