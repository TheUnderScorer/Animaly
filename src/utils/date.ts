import moment from 'moment';

export type TimeOfDay = 'night' | 'morning' | 'afternoon' | 'evening';

const timesOfDay = {
  '0-5': 'night',
  '6-10': 'morning',
  '11-15': 'afternoon',
  '16-23': 'evening',
};
const timesOfDayArr = Object.entries(timesOfDay);

export const getTimeOfDayText = (date = moment()): TimeOfDay | null => {
  const hour = parseInt(date.format('H'), 10);

  const foundTime = timesOfDayArr.find(([period]) => {
    const [startTime, endTime] = period
      .split('-')
      .map((item) => parseInt(item, 10));

    return hour >= startTime && hour <= endTime;
  });

  return foundTime ? (foundTime[1] as TimeOfDay) : null;
};
