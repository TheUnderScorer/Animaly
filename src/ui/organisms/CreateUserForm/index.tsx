import React, { FC, useCallback, useState } from 'react';
import { KeyboardAvoidingView, StyleSheet } from 'react-native';
import { Button, Icon, Input, Spinner, Text } from '@ui-kitten/components';
import { normalIcon } from '../../../styles/icons';
import { getBehavior } from '../../../utils/keyboardAvoidingView';
import { useDispatch } from 'react-redux';
import { createUser } from '../../../store/reducers/user/actions';
import { useAsyncStorageContext } from '../../../providers/AsyncStorageProvider';

export interface CreateUserFormProps {}

const styles = StyleSheet.create({
  input: {
    minWidth: 300,
  },
  container: {
    minWidth: 300,
  },
  title: {
    textAlign: 'center',
    marginBottom: 40,
  },
  btn: {
    marginTop: 40,
  },
  labelView: {
    alignItems: 'center',
  },
});

const CreateUserForm: FC<CreateUserFormProps> = () => {
  const { storage } = useAsyncStorageContext();
  const dispatch = useDispatch();

  const [loading, setLoading] = useState(false);
  const [name, setName] = useState('');
  const [error, setError] = useState('');

  const handleSave = useCallback(async () => {
    setError('');

    if (!name) {
      setError('Provide your name.');

      return;
    }

    setLoading(true);

    await dispatch(
      createUser(storage, {
        name,
      }),
    );
  }, [name, storage, dispatch]);

  return (
    <KeyboardAvoidingView behavior={getBehavior()} style={styles.container}>
      <Input
        disabled={loading}
        label={() => (
          <Text style={styles.title} category="h4">
            How should I call you?
          </Text>
        )}
        value={name}
        status={error ? 'danger' : 'basic'}
        size="large"
        caption={error ? error : 'It can be changed later.'}
        onChangeText={setName}
        placeholder="Enter your name..."
      />
      <Button
        disabled={loading}
        size="medium"
        onPress={handleSave}
        accessoryLeft={() =>
          loading ? (
            <Spinner />
          ) : (
            <Icon style={normalIcon} fill="#fff" name="save-outline" />
          )
        }
        style={styles.btn}>
        {loading ? '' : 'Save'}
      </Button>
    </KeyboardAvoidingView>
  );
};

export default CreateUserForm;
