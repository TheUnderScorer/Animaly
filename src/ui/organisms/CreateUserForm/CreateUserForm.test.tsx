import React, { ReactNode } from 'react';
import { Provider } from 'react-redux';
import store from '../../../store';
import { act, fireEvent, render } from 'react-native-testing-library';
import CreateUserForm from './index';
import { ApplicationProvider, IconRegistry } from '@ui-kitten/components';
import * as eva from '@eva-design/eva';
import AsyncStorageProvider from '../../../providers/AsyncStorageProvider';
import { EvaIconsPack } from '@ui-kitten/eva-icons';
import { wait } from '../../../utils/timeout';

const wrapComponent = (cmp: ReactNode) => (
  <ApplicationProvider {...eva} theme={eva.dark}>
    <IconRegistry icons={EvaIconsPack} />
    <AsyncStorageProvider>
      <Provider store={store}>{cmp}</Provider>
    </AsyncStorageProvider>
  </ApplicationProvider>
);

describe('<CreateUserForm />', () => {
  it('should render without crashing', () => {
    const component = render(wrapComponent(<CreateUserForm />));
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should show error if user leaves input empty', async () => {
    const component = render(wrapComponent(<CreateUserForm />));
    const btn = await component.findByTestId('createUserBtn');

    act(() => {
      fireEvent.press(btn);
    });

    const error = await component.findByText('Provide your name.');
    expect(error).toBeDefined();
  });

  it('should handle user creation', async () => {
    const handleCreation = jest.fn();

    const component = render(
      wrapComponent(<CreateUserForm onCreate={handleCreation} />),
    );
    const input = await component.findByTestId('createUserInput');

    act(() => {
      fireEvent.changeText(input, 'Test');
    });

    const btn = await component.findByTestId('createUserBtn');

    await act(async () => {
      fireEvent.press(btn);

      await wait(100);
    });

    expect(handleCreation).toHaveBeenCalledTimes(1);
  });
});
