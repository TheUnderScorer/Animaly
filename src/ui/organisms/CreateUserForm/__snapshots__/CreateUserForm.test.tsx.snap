// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateUserForm /> should render without crashing 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
  >
    <View
      style={
        Object {
          "bottom": 0,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "padding": 20,
          }
        }
      >
        <Text
          appearance="default"
          category="h3"
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "System",
                "fontSize": 30,
                "fontWeight": "800",
              },
              Object {
                "bottom": 40,
                "marginTop": 20,
                "textAlign": "center",
              },
            ]
          }
        >
          Welcome!
        </Text>
        <View
          style={
            Object {
              "height": 533.6,
              "width": 750,
            }
          }
        >
          <Image
            fadeDuration={1000}
            resizeMode="cover"
            source={
              Object {
                "testUri": "../../../src/assets/arabica-31.png",
              }
            }
            style={
              Object {
                "height": "100%",
                "width": "100%",
              }
            }
          />
          <Text
            appearance="default"
            category="p1"
            style={
              Array [
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "System",
                  "fontSize": 15,
                  "fontWeight": "400",
                },
                Object {
                  "bottom": "4%",
                  "paddingRight": 5,
                  "textAlign": "right",
                },
              ]
            }
          >
            illustration by 
            <Text
              appearance="default"
              category="p1"
              onPress={[Function]}
              style={
                Array [
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "System",
                    "fontSize": 15,
                    "fontWeight": "400",
                  },
                  Array [
                    Object {
                      "color": "blue",
                      "textDecorationColor": "black",
                      "textDecorationLine": "underline",
                      "textDecorationStyle": "solid",
                    },
                  ],
                ]
              }
            >
              Ouch.pics
            </Text>
          </Text>
        </View>
        <Text
          appearance="default"
          category="h5"
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "System",
                "fontSize": 22,
                "fontWeight": "800",
              },
              Object {
                "marginBottom": 50,
                "marginTop": 50,
                "textAlign": "center",
              },
            ]
          }
        >
          How should I call you?
        </Text>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "paddingLeft": 20,
                "paddingRight": 20,
                "width": "100%",
              },
            ]
          }
        >
          <View
            style={Object {}}
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#1A2138",
                    "borderColor": "#101426",
                    "borderRadius": 4,
                    "borderWidth": 1,
                    "minHeight": 48,
                    "paddingHorizontal": 8,
                    "paddingVertical": 11,
                  },
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "width": "100%",
                  },
                ]
              }
            >
              <TextInput
                allowFontScaling={true}
                appearance="default"
                disabled={false}
                editable={true}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                placeholder="Enter your name..."
                placeholderTextColor="#8F9BB3"
                rejectResponderTermination={true}
                size="large"
                status="basic"
                style={
                  Array [
                    Object {
                      "color": "#FFFFFF",
                      "fontFamily": "System",
                      "fontSize": 15,
                      "fontWeight": "normal",
                      "marginHorizontal": 8,
                    },
                    Object {
                      "flexBasis": "auto",
                      "flexGrow": 1,
                      "flexShrink": 1,
                    },
                    null,
                    undefined,
                  ]
                }
                testID="createUserInput"
                underlineColorAndroid="transparent"
                value=""
              />
            </View>
            <View
              style={
                Array [
                  Object {
                    "marginTop": 4,
                  },
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                ]
              }
            />
          </View>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#3366FF",
                "borderColor": "#3366FF",
                "borderRadius": 4,
                "borderWidth": 1,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginTop": 50,
                "minHeight": 32,
                "minWidth": 32,
                "opacity": 1,
                "paddingHorizontal": 8,
                "paddingVertical": 8,
                "width": "100%",
              }
            }
            testID="createUserBtn"
          >
            <Text
              appearance="default"
              category="p1"
              style={
                Array [
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "System",
                    "fontSize": 15,
                    "fontWeight": "400",
                  },
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "System",
                    "fontSize": 12,
                    "fontWeight": "bold",
                    "marginHorizontal": 8,
                  },
                ]
              }
            >
              Continue
            </Text>
            <View
              style={
                Object {
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                }
              }
            >
              <RNSVGSvgView
                align="xMidYMid"
                bbHeight={32}
                bbWidth={32}
                fill="#fff"
                focusable={false}
                meetOrSlice={0}
                minX={0}
                minY={0}
                style={
                  Array [
                    Object {
                      "backgroundColor": "transparent",
                      "borderWidth": 0,
                    },
                    Object {
                      "height": 32,
                      "width": 32,
                    },
                    Object {
                      "flex": 0,
                      "height": 32,
                      "width": 32,
                    },
                  ]
                }
                vbHeight={24}
                vbWidth={24}
              >
                <RNSVGGroup
                  fill={4294967295}
                  propList={
                    Array [
                      "fill",
                    ]
                  }
                >
                  <RNSVGGroup>
                    <RNSVGGroup>
                      <RNSVGRect
                        height="24"
                        matrix={
                          Array [
                            6.123233995736766e-17,
                            -1,
                            1,
                            6.123233995736766e-17,
                            0,
                            24,
                          ]
                        }
                        opacity={0}
                        width="24"
                        x={0}
                        y={0}
                      />
                      <RNSVGPath
                        d="M5 13h11.86l-3.63 4.36a1 1 0 0 0 1.54 1.28l5-6a1.19 1.19 0 0 0 .09-.15c0-.05.05-.08.07-.13A1 1 0 0 0 20 12a1 1 0 0 0-.07-.36c0-.05-.05-.08-.07-.13a1.19 1.19 0 0 0-.09-.15l-5-6A1 1 0 0 0 14 5a1 1 0 0 0-.64.23 1 1 0 0 0-.13 1.41L16.86 11H5a1 1 0 0 0 0 2z"
                      />
                    </RNSVGGroup>
                  </RNSVGGroup>
                </RNSVGGroup>
              </RNSVGSvgView>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
